src/View/Shortcodes/EmbedShortcodeProvider.php

```diff
@@ -84,32 +86,27 @@ public static function handle_shortcode($arguments, $content, $parser, $shortcod
             $serviceArguments['min_image_height'] = $arguments['height'];
         }
 
-        /** @var EmbedResource $embed */
-        $embed = Injector::inst()->create(Embeddable::class, $serviceURL);
-        if (!empty($serviceArguments)) {
-            $embed->setOptions(array_merge($serviceArguments, (array) $embed->getOptions()));
+        /** @var Embeddable $embeddable */
+        // Allow mocking of Embeddable
+        if (($extra['Embeddable'] ?? null) instanceof Embeddable) {
+            $embeddable = $extra['Embeddable'];
+        } else {
+            $embeddable = Injector::inst()->create(Embeddable::class, $serviceURL);
         }
-
-        // Allow resolver to be mocked
-        $dispatcher = null;
-        if (isset($extra['resolver'])) {
-            $dispatcher = Injector::inst()->create(
-                $extra['resolver']['class'],
-                $serviceURL,
-                $extra['resolver']['config']
-            );
-        } elseif (Injector::inst()->has(DispatcherInterface::class)) {
-            $dispatcher = Injector::inst()->get(DispatcherInterface::class);
+        if (!empty($serviceArguments)) {
+            $embeddable->setOptions(array_merge($serviceArguments, (array) $embeddable->getOptions()));
         }
 
-        if ($dispatcher) {
-            $embed->setDispatcher($dispatcher);
+        // Only EmbedContainer is currently supported
+        if (!($embeddable instanceof EmbedContainer)) {
+            throw new Exception('Emeddable must extend EmbedContainer');
```

Comment:
`$embeddable->getExtractor()` is stopping us

The concept of `extractor` is specific to embed v4.  embed v3 has a different concept 'adapter'

The reason we're using an interface here is because it was already there, not because there's an expectation that we'll be supporting more implementing class.  The interface does admittedly make it easier to make it easier to swap out to the newer implementation. 

What's the possibility that someone wants a different Embeddable here?  It implies supporting multiple major versions of embed/embed.  For instance we could in theory keep supporting embed/embed v3 (which no longer receives updates), change the requirement to "embed/embed": "^3 || ^4", so we do further work to make EmbedResource (v3) and EmbedContainer (v4) both work.

But I think this would be a poor investment of effort

Reply:
Updated